lib/Graph/Graph.cpp

```diff
@@ -1482,6 +1482,58 @@ Node *Function::createWeightedSum(llvm::StringRef name,
   return currAdd;
 }
 
+Node *Function::createBatchBoxCox(llvm::StringRef name, NodeValue data,
+                                  NodeValue lambda1, NodeValue lambda2) {
+  assert((lambda1.dims() == lambda2.dims()) &&
+         "lambda1 and lambda2 must have the same shape");
+  assert((lambda1.getType()->getElementType() == lambda2.getElementType()) &&
+         "lambda1 and lambda2 must have the same element type");
+  assert((lambda1.getType()->getElementType() == data.getElementType()) &&
+         "data and lambdas must have the same element type");
+  assert((lambda1.dims().size() == 1) && "lambda1 and lambda2 must be vectors");
+  assert((data.dims().size() == 2) && "data must be a matrix");
+  assert((data.dims()[1] == lambda1.dims()[0]) &&
+         "data, lambda1 and lambda2 must have the same number of rows");
+
+  // Broadcast lambda1 and lambda2 so that they are both the same size as the
+  // data.
+  auto *BL1 = createBroadcast(name.str() + ".broadcast", lambda1, data.dims(),
+                              /*axis=*/1);
+  auto *BL2 = createBroadcast(name.str() + ".broadcast", lambda2, data.dims(),
+                              /*axis=*/1);
+
+  // Compute data + BL2, which is needed regardless of whether
+  // lambda1 is 0 or not.
+  auto *AN = createAdd(name.str() + ".add", data, BL2);
+
+  // Compute the Box-Cox transform for the lambda1 == 0 case:
+  //    y = ln(x + lambda2)
+
+  // Clip the data to the log below at 1e-6 to avoid taking the logarithm
+  // of too small a number.
+  auto *minArg =
+      createSplat(name.str() + ".logminarg", AN->getResult().getType(), 1e-6);
+  auto *MN = createMax(name.str() + ".logmax", AN, minArg);
+  auto *LN = createLog(name.str() + ".log", MN);
+
+  // Compute the Box-Cox transform for the lambda1 != 0 case:
+  //    y = ((x + lambda2)^lambda1 - 1)/lambda1
+  auto *PN = createPow(name.str() + ".pow", AN, BL1);
+  auto *ones =
+      createSplat(name.str() + ".ones", PN->getResult().getType(), 1.0f);
+  auto *SN = createSub(name.str() + ".sub", PN, ones);
+  auto *DN = createDiv(name.str() + ".div", SN, BL1);
```

Comment:
Unfortunately this is going to execute for all elements regardless of if BL1 is zero. So it's perhaps the source of your UBSAN div by zero error. The way you have implemented this, both paths are executed and one is just selected in the end. This also means we are doing unnecessary work. It would be a great place for us to use predication, but of course it's not supported yet.

For now you could do something like CmpEQ(0) + Select a Splat close to zero as lambda1 to avoid div by zero, even though you know it will be thrown away in the final select regardless. It's not ideal though, as you need another Select... Anyone else have other ideas?

Reply:
> It would be a great place for us to use predication, but of course it's not supported yet.

Predication as we envision it is not required for correctness. (Some hardwares may not support it.) Therefore we need to make sure we don't execute division by zero.

> For now you could do something like CmpEQ(0) + Select a Splat close to zero as lambda1 to avoid div by zero, even though you know it will be thrown away in the final select regardless. It's not ideal though, as you need another Select... Anyone else have other ideas?

If we are going to throw the result away, I'd rather divide by one or anything that is likely to be fast to execute. Could we do the computation on only what make sense instead?
(I haven't read the patch nor the semantic of the operator.)

Reply:
Okay read the operator semantic and this "Could we do the computation on only what make sense instead?" doesn't seem to be practical.
So yeah, prepare the divider upfront sounds reasonable to me.

Reply:
Assuming this doesn't affect the precision too much we could avoid the select with a splat and just add epsilon to every value.
I.e., instead of
```
predicate = input == 0
divider = select predicate, input, splat(1)
```
We could do
```
divider = input + splat(epsilon)
```
(This assumes that the input is not -epsilon though...)